Attachment support

ringtailsoftware opened the pull request:
I've added support for attaching a single file.
It was working here (with my other changes), until I merged with the latest composer changes.

Raising a PR in case it's obvious to you what's gone wrong, else I'll continue later...

ringtailsoftware committed:
Add file attachment support.
Only supports a single file. Doesn't generate a blurhash, so Mastodon seems to show the image as "Not available" for a few seconds.

ringtailsoftware committed:
Fix posting with no attachment

benbrown commented on index.js:
```diff
@@ -73,7 +73,8 @@ app.use(bodyParser.json({
 app.use(bodyParser.json({
-  type: 'application/json'
+  type: 'application/json',
+  limit: '4mb' // allow large bodies as attachments are base64 in JSON
```
seems low.

benbrown commented on lib/account.js:
```diff
@@ -25,7 +25,9 @@ import {
   getFileName,
   boostsFile,
-  pathToDMs
+  pathToDMs,
```
this is not used

ringtailsoftware replied:
Not sure what you mean here, wrong line?

benbrown replied:
nevermind i misread the diff.

benbrown commented on lib/account.js:
```diff
@@ -118,6 +120,16 @@ export const acceptDM = (dm, inboxUser) => {
 
 }
 
+export const writeMedia = (filename, attachment) => {
```
We should do the processing inside the main writeMediaFile/readMediaFile functions in storage.js to make it transparent here.

benbrown commented on public/app.js:
```diff
@@ -162,11 +162,34 @@ const app = {
     },
-    post: () => {
+    readAttachment: async () => {
+        // read the file into base64, return mimtype and data
```
typo mimtype -> mimetype

benbrown commented:
This is super cool, thank you for this contribution.

One thing I am wondering is if there is a very strong benefit to wrapping the media as base64 in a json file?  Does it make it easier to handle in express?

My concerns are that his will result in large files that can't be easily inspected. I would prefer to write them as normal files so that the owner of the instance can look at them without extra tooling. Also, this would allow them to be served out of a _static_ media folder instead of processing them through a handler.

Using the hash of the content to create the filename is nice since it elegantly handles duplicate files.

Also, this will need a way to add the content description as well -- I am happy to work on that if you would prefer?

benbrown reviewed (changes requested):
Thanks again! See comments.

benbrown commented:
In my testing, the POST will fail for files that are larger than about 1 MB. Chrome debugger doesn't really show much - it says the request canceled.

I *suspect* this has to do with Express hitting the payload size, but I'm not sure. Base64 stuff gets big fast.

ringtailsoftware committed:
app.post() cannot be async, as onsubmit calls it synchronously. Instead, return immediately and handle the rest in .then()

This was broken on Firefox, seemed ok in Chrome.

ringtailsoftware committed:
Fix typo

ringtailsoftware commented:
I agree about base64 encoded files getting big. I only chose to do it that way so that each individual media file on disk would have the binary data and the metadata (mimetype, + description, blurhash etc in future).
If I just wrote the raw datafile to disk (named by hash), is there already a good place to stash the metadata?
Or, should I have a <hash>_meta.json file alongside?

ringtailsoftware committed:
Increase express max payload

ringtailsoftware committed:
Move file structure logic out of storage.js

ringtailsoftware committed:
For media, store 2 files. <hash> for the binary data, <hash>.json for the metadata

ringtailsoftware committed:
Add alt-text (description) for media uploads

ringtailsoftware committed:
Add blurhash and image dimensions support
Fix bug where sensitive was enabled by default (cw == undefined, not null)

ringtailsoftware commented:
> In my testing, the POST will fail for files that are larger than about 1 MB. Chrome debugger doesn't really show much - it says the request canceled.
> 
> I _suspect_ this has to do with Express hitting the payload size, but I'm not sure. Base64 stuff gets big fast.

I've just tested ok with an 8MB PNG (Firefox). Took a little while but worked. Initially it was failing due to nginx (https://www.cyberciti.biz/faq/linux-unix-bsd-nginx-413-request-entity-too-large/)

ringtailsoftware commented:
@benbrown I'm done. Ready for you to try it.

benbrown commented:
you are on fire @ringtailsoftware. I was just looking at your recent commits. I'll try it out when I can!!

benbrown commented:
Why do we need to store the metadata separately from where it lives in the post itself?

I think we get mime-type for free if the file gets written out to /media/hash.png

benbrown commented:
@ringtailsoftware this is awesome. I might put some work into this while you sleep -- if the power doesn't go out in texas!!

ringtailsoftware commented:
> 

If we use the file extension then we'd still have to decode ".png" into "image/png" and so on.
In theory, so long as the browser sends a valid mimetype this scheme should work for video and other formats without a mapping table.

benbrown commented:
The original filename must be available in the browser - it could be passed as part of the payload, and then can use that to apply the extension.  Ultimately I think it is important to have human readable files in the /media folder.

ringtailsoftware commented:
> The original filename must be available in the browser - it could be passed as part of the payload, and then can use that to apply the extension. Ultimately I think it is important to have human readable files in the /media folder (meaning I can look at them with my normal tools)

Ok, in that case, I'd suggest taking the mimetype "image/png" and renaming the data file to `${hash}.${mimetype.split('/')[1]}`. That way the original filename doesn't matter, but you can still view the files easily.

I'm concerned about someone uploading a file called "../../" or a PNG called "foo.jpg", etc.

ringtailsoftware committed:
Use mimetype to choose meaningful filename for data file

ringtailsoftware commented:
That'll do it

ringtailsoftware committed:
In the ActivityPub Note, set the URL of the media file to <hash>.ext (eg. abc123.png). This means that the media directory could be served by a static webserver.
When a request comes in via existing /media/:id endpoint, drop the .ext and lookup the mime type (for Content-Type) via the metadata.

ringtailsoftware commented:
Minor tweak made to allow media directory to be statically served
